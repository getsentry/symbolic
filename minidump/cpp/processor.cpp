#include "google_breakpad/processor/basic_source_line_resolver.h"
#include "google_breakpad/processor/minidump.h"
#include "google_breakpad/processor/minidump_processor.h"
#include "google_breakpad/processor/process_state.h"

#include "cpp/data_definitions.h"
#include "cpp/memstream.h"
#include "cpp/mmap_symbol_supplier.h"
#include "cpp/processor.h"

using google_breakpad::BasicSourceLineResolver;
using google_breakpad::Minidump;
using google_breakpad::MinidumpProcessor;
using google_breakpad::ProcessState;

process_state_t *process_minidump(const char *buffer,
                                  size_t buffer_size,
                                  symbol_entry_t *symbols,
                                  size_t symbol_count,
                                  int *result_out) {
    if (buffer == nullptr) {
        *result_out = google_breakpad::PROCESS_ERROR_MINIDUMP_NOT_FOUND;
        return nullptr;
    }

    ProcessState *state = new ProcessState();
    if (state == nullptr) {
        *result_out = -1;  // Memory allocation issue
        return nullptr;
    }

    BasicSourceLineResolver resolver;
    MmapSymbolSupplier supplier(symbol_count, symbols);
    MinidumpProcessor processor(&supplier, &resolver);

    imemstream in(buffer, buffer_size);
    Minidump minidump(in);
    if (!minidump.Read()) {
        *result_out = google_breakpad::PROCESS_ERROR_MINIDUMP_NOT_FOUND;
        return nullptr;
    }

    *result_out = processor.Process(&minidump, state);
    if (*result_out != google_breakpad::PROCESS_OK) {
        delete state;
        return nullptr;
    }

    return process_state_t::cast(state);
}
