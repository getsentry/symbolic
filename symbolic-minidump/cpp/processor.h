#ifndef SENTRY_PROCESSOR_H
#define SENTRY_PROCESSOR_H

#include <cstddef>
#include "cpp/data_structures.h"

#ifdef __cplusplus
extern "C" {
#endif

/// Data transfer object for symbols in memory
struct symbol_entry_t {
    /// The debug identifier of the code module these symbols are for
    const char *debug_identifier;

    /// Size of the buffer inside symbol_data
    const size_t symbol_size;

    /// Raw data of the symbol file passed to the symbolizer
    const char *symbol_data;
};

/// Reads a minidump from a memory buffer and processes it. Returns an owning
/// pointer to a process_state_t struct that contains loaded code modules and
/// call stacks of all threads of the process during the crash.
///
/// Processing the minidump can fail if the buffer is corrupted or does not
/// exit. The function will return NULL and an error code in result_out.
///
/// Release memory of the process state with process_state_delete.
process_state_t *process_minidump(const char *buffer,
                                  size_t buffer_size,
                                  symbol_entry_t *symbols,
                                  size_t symbol_count,
                                  int *result_out);

#ifdef __cplusplus
}
#endif

#endif
