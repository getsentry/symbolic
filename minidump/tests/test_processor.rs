use failure::Error;
use insta;

use symbolic_common::ByteView;
use symbolic_minidump::processor::ProcessState;

#[test]
fn process_minidump_linux() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/linux/mini.dmp")?;
    let state = ProcessState::from_minidump(&buffer, None)?;
    insta::assert_debug_snapshot_matches!("process_state_linux", &state);
    Ok(())
}

#[test]
fn process_minidump_macos() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/macos/mini.dmp")?;
    let state = ProcessState::from_minidump(&buffer, None)?;
    insta::assert_debug_snapshot_matches!("process_state_macos", &state);
    Ok(())
}

#[test]
fn process_minidump_windows() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/windows/mini.dmp")?;
    let state = ProcessState::from_minidump(&buffer, None)?;
    insta::assert_debug_snapshot_matches!("process_state_windows", &state);
    Ok(())
}

#[test]
fn get_referenced_modules_linux() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/linux/mini.dmp")?;
    let state = ProcessState::from_minidump(&buffer, None)?;
    insta::assert_debug_snapshot_matches!("referenced_modules_linux", &state.referenced_modules());
    Ok(())
}

#[test]
fn get_referenced_modules_macos() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/macos/mini.dmp")?;
    let state = ProcessState::from_minidump(&buffer, None)?;
    insta::assert_debug_snapshot_matches!("referenced_modules_macos", &state.referenced_modules());
    Ok(())
}

#[test]
fn get_referenced_modules_windows() -> Result<(), Error> {
    let buffer = ByteView::open("../testutils/fixtures/windows/mini.dmp")?;
    let state = ProcessState::from_minidump(&buffer, None)?;
    insta::assert_debug_snapshot_matches!(
        "referenced_modules_windows",
        &state.referenced_modules()
    );
    Ok(())
}
