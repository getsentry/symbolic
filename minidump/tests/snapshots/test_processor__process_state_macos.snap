---
created: "2019-03-20T08:50:27.479637Z"
creator: insta@0.7.1
source: minidump/tests/test_processor.rs
expression: "&state"
---
ProcessState {
    requesting_thread: 0,
    timestamp: 1521713398,
    crash_address: 69,
    crash_reason: "EXC_BAD_ACCESS / KERN_INVALID_ADDRESS",
    assertion: "",
    system_info: SystemInfo {
        os_name: "Mac OS X",
        os_version: "10.12.6",
        cpu_family: "amd64",
        cpu_info: "family 6 model 58 stepping 9",
        cpu_count: 2
    },
    threads: [
        CallStack {
            thread_id: 775,
            frames: [
                StackFrame {
                    return_address: 4458187797,
                    instruction: 4458187797,
                    trust: Context,
                    module: Some(
                        CodeModule {
                            id: Some(
                                CodeModuleId {
                                    inner: DebugId {
                                        uuid: "67e9247c-814e-392b-a027-dbde6748fcbf",
                                        appendix: 0
                                    }
                                }
                            ),
                            base_address: 4458131456,
                            size: 69632,
                            code_file: "/Users/travis/build/getsentry/breakpad-tools/macos/build/./crash",
                            code_identifier: "",
                            debug_file: "crash",
                            debug_identifier: "67E9247C814E392BA027DBDE6748FCBF0"
                        }
                    )
                },
                StackFrame {
                    return_address: 4458187888,
                    instruction: 4458187887,
                    trust: Scan,
                    module: Some(
                        CodeModule {
                            id: Some(
                                CodeModuleId {
                                    inner: DebugId {
                                        uuid: "67e9247c-814e-392b-a027-dbde6748fcbf",
                                        appendix: 0
                                    }
                                }
                            ),
                            base_address: 4458131456,
                            size: 69632,
                            code_file: "/Users/travis/build/getsentry/breakpad-tools/macos/build/./crash",
                            code_identifier: "",
                            debug_file: "crash",
                            debug_identifier: "67E9247C814E392BA027DBDE6748FCBF0"
                        }
                    )
                },
                StackFrame {
                    return_address: 140737084568117,
                    instruction: 140737084568116,
                    trust: Scan,
                    module: Some(
                        CodeModule {
                            id: Some(
                                CodeModuleId {
                                    inner: DebugId {
                                        uuid: "9b2ac56d-107c-3541-a127-9094a751f2c9",
                                        appendix: 0
                                    }
                                }
                            ),
                            base_address: 140737084547072,
                            size: 24576,
                            code_file: "/usr/lib/system/libdyld.dylib",
                            code_identifier: "",
                            debug_file: "libdyld.dylib",
                            debug_identifier: "9B2AC56D107C3541A1279094A751F2C90"
                        }
                    )
                },
                StackFrame {
                    return_address: 140737084568117,
                    instruction: 140737084568116,
                    trust: Scan,
                    module: Some(
                        CodeModule {
                            id: Some(
                                CodeModuleId {
                                    inner: DebugId {
                                        uuid: "9b2ac56d-107c-3541-a127-9094a751f2c9",
                                        appendix: 0
                                    }
                                }
                            ),
                            base_address: 140737084547072,
                            size: 24576,
                            code_file: "/usr/lib/system/libdyld.dylib",
                            code_identifier: "",
                            debug_file: "libdyld.dylib",
                            debug_identifier: "9B2AC56D107C3541A1279094A751F2C90"
                        }
                    )
                }
            ]
        }
    ],
    modules: [
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "67e9247c-814e-392b-a027-dbde6748fcbf",
                        appendix: 0
                    }
                }
            ),
            base_address: 4458131456,
            size: 69632,
            code_file: "/Users/travis/build/getsentry/breakpad-tools/macos/build/./crash",
            code_identifier: "",
            debug_file: "crash",
            debug_identifier: "67E9247C814E392BA027DBDE6748FCBF0"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "36385a3a-60d3-32db-bf55-c6d8931a7aa6",
                        appendix: 0
                    }
                }
            ),
            base_address: 140736719339520,
            size: 4800512,
            code_file: "/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation",
            code_identifier: "",
            debug_file: "CoreFoundation",
            debug_identifier: "36385A3A60D332DBBF55C6D8931A7AA60"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "84a04d24-0e60-3810-a8c0-90a65e2df61a",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737059020800,
            size: 8192,
            code_file: "/usr/lib/libDiagnosticMessagesClient.dylib",
            code_identifier: "",
            debug_file: "libDiagnosticMessagesClient.dylib",
            debug_identifier: "84A04D240E603810A8C090A65E2DF61A0"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "f18ac1e7-c6f1-34b1-8069-be571b3231d4",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737061376000,
            size: 8192,
            code_file: "/usr/lib/libSystem.B.dylib",
            code_identifier: "",
            debug_file: "libSystem.B.dylib",
            debug_identifier: "F18AC1E7C6F134B18069BE571B3231D40"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "0b43bb5d-e6eb-3464-8de9-b41ac8ed9d1c",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737063157760,
            size: 356352,
            code_file: "/usr/lib/libc++.1.dylib",
            code_identifier: "",
            debug_file: "libc++.1.dylib",
            debug_identifier: "0B43BB5DE6EB34648DE9B41AC8ED9D1C0"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "bc271ad3-831b-362a-9da7-e8c51f285fe4",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737063514112,
            size: 172032,
            code_file: "/usr/lib/libc++abi.dylib",
            code_identifier: "",
            debug_file: "libc++abi.dylib",
            debug_identifier: "BC271AD3831B362A9DA7E8C51F285FE40"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "ccd2ed24-3071-383b-925d-8d763bb12a6f",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737069191168,
            size: 2252800,
            code_file: "/usr/lib/libicucore.A.dylib",
            code_identifier: "",
            debug_file: "libicucore.A.dylib",
            debug_identifier: "CCD2ED243071383B925D8D763BB12A6F0"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "4df3c25c-52c2-3f01-a3ef-0d9d53a73c1c",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737075171328,
            size: 4022272,
            code_file: "/usr/lib/libobjc.A.dylib",
            code_identifier: "",
            debug_file: "libobjc.A.dylib",
            debug_identifier: "4DF3C25C52C23F01A3EF0D9D53A73C1C0"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "46e3ffa2-4328-327a-8d34-a03e20bffb8e",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737083535360,
            size: 73728,
            code_file: "/usr/lib/libz.1.dylib",
            code_identifier: "",
            debug_file: "libz.1.dylib",
            debug_identifier: "46E3FFA24328327A8D34A03E20BFFB8E0"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "093a4dab-8385-3d47-a350-e20cb7ccf7bf",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737083666432,
            size: 20480,
            code_file: "/usr/lib/system/libcache.dylib",
            code_identifier: "",
            debug_file: "libcache.dylib",
            debug_identifier: "093A4DAB83853D47A350E20CB7CCF7BF0"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "8a64d1b0-c70e-385c-92f0-e669079fda90",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737083686912,
            size: 45056,
            code_file: "/usr/lib/system/libcommonCrypto.dylib",
            code_identifier: "",
            debug_file: "libcommonCrypto.dylib",
            debug_identifier: "8A64D1B0C70E385C92F0E669079FDA900"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "55d47421-772a-32ab-b529-1a46c2f43b4d",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737083731968,
            size: 32768,
            code_file: "/usr/lib/system/libcompiler_rt.dylib",
            code_identifier: "",
            debug_file: "libcompiler_rt.dylib",
            debug_identifier: "55D47421772A32ABB5291A46C2F43B4D0"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "819bea3c-df11-3e3d-a1a1-5a51c5bf1961",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737083764736,
            size: 36864,
            code_file: "/usr/lib/system/libcopyfile.dylib",
            code_identifier: "",
            debug_file: "libcopyfile.dylib",
            debug_identifier: "819BEA3CDF113E3DA1A15A51C5BF19610"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "65d7165e-2e71-335d-a2d6-33f78e2df0c1",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737083801600,
            size: 540672,
            code_file: "/usr/lib/system/libcorecrypto.dylib",
            code_identifier: "",
            debug_file: "libcorecrypto.dylib",
            debug_identifier: "65D7165E2E71335DA2D633F78E2DF0C10"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "6582bad6-ed27-3b30-b620-90b1c5a4ae3c",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737084342272,
            size: 204800,
            code_file: "/usr/lib/system/libdispatch.dylib",
            code_identifier: "",
            debug_file: "libdispatch.dylib",
            debug_identifier: "6582BAD6ED273B30B62090B1C5A4AE3C0"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "9b2ac56d-107c-3541-a127-9094a751f2c9",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737084547072,
            size: 24576,
            code_file: "/usr/lib/system/libdyld.dylib",
            code_identifier: "",
            debug_file: "libdyld.dylib",
            debug_identifier: "9B2AC56D107C3541A1279094A751F2C90"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "7aa011a9-dc21-3488-bf73-3b5b14d1fdd6",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737084571648,
            size: 4096,
            code_file: "/usr/lib/system/libkeymgr.dylib",
            code_identifier: "",
            debug_file: "libkeymgr.dylib",
            debug_identifier: "7AA011A9DC213488BF733B5B14D1FDD60"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "b856abd2-896e-3de0-b2c8-146a6af8e2a7",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737084628992,
            size: 4096,
            code_file: "/usr/lib/system/liblaunch.dylib",
            code_identifier: "",
            debug_file: "liblaunch.dylib",
            debug_identifier: "B856ABD2896E3DE0B2C8146A6AF8E2A70"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "17d5d855-f6c3-3b04-b680-e9bf02ef8aed",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737084633088,
            size: 24576,
            code_file: "/usr/lib/system/libmacho.dylib",
            code_identifier: "",
            debug_file: "libmacho.dylib",
            debug_identifier: "17D5D855F6C33B04B680E9BF02EF8AED0"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "12448cc2-378e-35f3-be33-9dc395a5b970",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737084657664,
            size: 12288,
            code_file: "/usr/lib/system/libquarantine.dylib",
            code_identifier: "",
            debug_file: "libquarantine.dylib",
            debug_identifier: "12448CC2378E35F3BE339DC395A5B9700"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "38d4cb9c-10cd-30d3-8b7b-a515ec75fe85",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737084669952,
            size: 8192,
            code_file: "/usr/lib/system/libremovefile.dylib",
            code_identifier: "",
            debug_file: "libremovefile.dylib",
            debug_identifier: "38D4CB9C10CD30D38B7BA515EC75FE850"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "096e4228-3b7c-30a6-8b13-ec909a64499a",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737084678144,
            size: 102400,
            code_file: "/usr/lib/system/libsystem_asl.dylib",
            code_identifier: "",
            debug_file: "libsystem_asl.dylib",
            debug_identifier: "096E42283B7C30A68B13EC909A64499A0"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "10dc5404-73ab-35b3-a277-a8afecb476eb",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737084780544,
            size: 4096,
            code_file: "/usr/lib/system/libsystem_blocks.dylib",
            code_identifier: "",
            debug_file: "libsystem_blocks.dylib",
            debug_identifier: "10DC540473AB35B3A277A8AFECB476EB0"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "e5ae5244-7d0c-36ac-8bb6-c7ae7ea52a4b",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737084784640,
            size: 581632,
            code_file: "/usr/lib/system/libsystem_c.dylib",
            code_identifier: "",
            debug_file: "libsystem_c.dylib",
            debug_identifier: "E5AE52447D0C36AC8BB6C7AE7EA52A4B0"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "becc01a2-ca8d-31e6-bcdf-d452965fa976",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737085366272,
            size: 16384,
            code_file: "/usr/lib/system/libsystem_configuration.dylib",
            code_identifier: "",
            debug_file: "libsystem_configuration.dylib",
            debug_identifier: "BECC01A2CA8D31E6BCDFD452965FA9760"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "7d26de79-b424-3450-85e1-f7fab32714ab",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737085382656,
            size: 16384,
            code_file: "/usr/lib/system/libsystem_coreservices.dylib",
            code_identifier: "",
            debug_file: "libsystem_coreservices.dylib",
            debug_identifier: "7D26DE79B424345085E1F7FAB32714AB0"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "ec6fcf07-dcfb-3a03-9cc9-6dd3709974c6",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737085399040,
            size: 102400,
            code_file: "/usr/lib/system/libsystem_coretls.dylib",
            code_identifier: "",
            debug_file: "libsystem_coretls.dylib",
            debug_identifier: "EC6FCF07DCFB3A039CC96DD3709974C60"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "cc960215-0b1b-3822-a13a-3dde96fa796f",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737085501440,
            size: 28672,
            code_file: "/usr/lib/system/libsystem_dnssd.dylib",
            code_identifier: "",
            debug_file: "libsystem_dnssd.dylib",
            debug_identifier: "CC9602150B1B3822A13A3DDE96FA796F0"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "611db84c-bf70-3f92-8702-b9f28a900920",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737085530112,
            size: 172032,
            code_file: "/usr/lib/system/libsystem_info.dylib",
            code_identifier: "",
            debug_file: "libsystem_info.dylib",
            debug_identifier: "611DB84CBF703F928702B9F28A9009200"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "34b1f16c-bc9c-3c5f-9045-0cae91cb5914",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737085702144,
            size: 143360,
            code_file: "/usr/lib/system/libsystem_kernel.dylib",
            code_identifier: "",
            debug_file: "libsystem_kernel.dylib",
            debug_identifier: "34B1F16CBC9C3C5F90450CAE91CB59140"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "86d499b5-bbdc-3d3b-8a4e-97ae8e6672a4",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737085845504,
            size: 294912,
            code_file: "/usr/lib/system/libsystem_m.dylib",
            code_identifier: "",
            debug_file: "libsystem_m.dylib",
            debug_identifier: "86D499B5BBDC3D3B8A4E97AE8E6672A40"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "a3d15f17-99a6-3367-8c7e-4280e8619c95",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737086140416,
            size: 126976,
            code_file: "/usr/lib/system/libsystem_malloc.dylib",
            code_identifier: "",
            debug_file: "libsystem_malloc.dylib",
            debug_identifier: "A3D15F1799A633678C7E4280E8619C950"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "369d0221-56ca-3c3e-9ede-94b41cae77b7",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737086267392,
            size: 368640,
            code_file: "/usr/lib/system/libsystem_network.dylib",
            code_identifier: "",
            debug_file: "libsystem_network.dylib",
            debug_identifier: "369D022156CA3C3E9EDE94B41CAE77B70"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "b021f2b3-8a75-3633-abb0-fc012b8e9b0c",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737086636032,
            size: 40960,
            code_file: "/usr/lib/system/libsystem_networkextension.dylib",
            code_identifier: "",
            debug_file: "libsystem_networkextension.dylib",
            debug_identifier: "B021F2B38A753633ABB0FC012B8E9B0C0"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "b8160190-a069-3b3a-bdf6-2aa408221fae",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737086676992,
            size: 40960,
            code_file: "/usr/lib/system/libsystem_notify.dylib",
            code_identifier: "",
            debug_file: "libsystem_notify.dylib",
            debug_identifier: "B8160190A0693B3ABDF62AA408221FAE0"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "897462fd-b318-321b-a554-e61982630f7e",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737086717952,
            size: 36864,
            code_file: "/usr/lib/system/libsystem_platform.dylib",
            code_identifier: "",
            debug_file: "libsystem_platform.dylib",
            debug_identifier: "897462FDB318321BA554E61982630F7E0"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "b8fb5e20-3295-39e2-b5eb-b464d1d4b104",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737086754816,
            size: 45056,
            code_file: "/usr/lib/system/libsystem_pthread.dylib",
            code_identifier: "",
            debug_file: "libsystem_pthread.dylib",
            debug_identifier: "B8FB5E20329539E2B5EBB464D1D4B1040"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "4b92ec49-acd0-36ae-b07a-a2b8152eaf9d",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737086799872,
            size: 16384,
            code_file: "/usr/lib/system/libsystem_sandbox.dylib",
            code_identifier: "",
            debug_file: "libsystem_sandbox.dylib",
            debug_identifier: "4B92EC49ACD036AEB07AA2B8152EAF9D0"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "f78b847b-3565-3e4b-98a6-f7ad40392e2d",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737086816256,
            size: 8192,
            code_file: "/usr/lib/system/libsystem_secinit.dylib",
            code_identifier: "",
            debug_file: "libsystem_secinit.dylib",
            debug_identifier: "F78B847B35653E4B98A6F7AD40392E2D0"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "3390e07c-c1ce-348f-adbd-2c5440b45eaa",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737086824448,
            size: 32768,
            code_file: "/usr/lib/system/libsystem_symptoms.dylib",
            code_identifier: "",
            debug_file: "libsystem_symptoms.dylib",
            debug_identifier: "3390E07CC1CE348FADBD2C5440B45EAA0"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "ac63a7fe-50d9-3a30-96e6-f6b7ff16e465",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737086857216,
            size: 81920,
            code_file: "/usr/lib/system/libsystem_trace.dylib",
            code_identifier: "",
            debug_file: "libsystem_trace.dylib",
            debug_identifier: "AC63A7FE50D93A3096E6F6B7FF16E4650"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "3d50d8a8-c460-334d-a519-2da841102c6b",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737086939136,
            size: 24576,
            code_file: "/usr/lib/system/libunwind.dylib",
            code_identifier: "",
            debug_file: "libunwind.dylib",
            debug_identifier: "3D50D8A8C460334DA5192DA841102C6B0"
        },
        CodeModule {
            id: Some(
                CodeModuleId {
                    inner: DebugId {
                        uuid: "bf896df0-d8e9-31a8-a4b3-01120bfeee52",
                        appendix: 0
                    }
                }
            ),
            base_address: 140737086963712,
            size: 172032,
            code_file: "/usr/lib/system/libxpc.dylib",
            code_identifier: "",
            debug_file: "libxpc.dylib",
            debug_identifier: "BF896DF0D8E931A8A4B301120BFEEE520"
        }
    ]
}
